third_party/libunwindstack/tests/PeCoffEpilogTest.cpp

```diff
@@ -0,0 +1,579 @@
+        machine_code->emplace_back(modrm);
+        std::array<uint8_t, 4> values;
+        uint32_t lea_displacement = options.lea_displacement;
+        std::memcpy(static_cast<void*>(&values[0]), static_cast<void*>(&lea_displacement),
+                    sizeof(uint32_t));
+        machine_code->insert(machine_code->end(), values.begin(), values.end());
+      }
+      expected_stack_pointer_after_unwind_ = options.rbp_value + options.lea_displacement;
+    }
+    if (options.insert_add_instruction) {
+      CHECK(options.added_value > 0);
+      machine_code->emplace_back(0x48);
+      if (options.added_value <= 0xff) {
+        machine_code->insert(machine_code->end(), {0x83, 0xc4});
+        uint8_t immediate = static_cast<uint8_t>(options.added_value);
+        machine_code->emplace_back(immediate);
+      } else {
+        machine_code->insert(machine_code->end(), {0x81, 0xc4});
+        std::array<uint8_t, 4> values;
+        uint32_t added_value = options.added_value;
+        std::memcpy(static_cast<void*>(&values[0]), static_cast<void*>(&added_value),
+                    sizeof(uint32_t));
+        machine_code->insert(machine_code->end(), values.begin(), values.end());
+      }
+      expected_stack_pointer_after_unwind_ += options.added_value;
+    }
+
+    for (const auto& pop_op : options.pop_operations) {
+      process_mem_fake_->SetData64(expected_stack_pointer_after_unwind_, pop_op.value);
+      expected_stack_pointer_after_unwind_ += sizeof(uint64_t);
+      // For 'pop' operations, the REX prefix is only used when one of R8 to R15 is
+      // the operand. In that case we need to use the fixed value of 0x41 as the
+      // REX prefix.
+      if (pop_op.reg >= R8) {
+        machine_code->emplace_back(0x41);
+      }
+      // Only want the lower three bits of the register value. If the highest bit is
+      // 1, this is indicated by presence of the REX prefix.
+      uint8_t opcode_byte = 0x58 | (pop_op.reg & 0x7);
+      machine_code->emplace_back(opcode_byte);
+    }
+
+    process_mem_fake_->SetData64(expected_stack_pointer_after_unwind_, options.return_address);
+    expected_stack_pointer_after_unwind_ += sizeof(uint64_t);
+
+    CHECK(options.jmp_instruction_bytes.empty() != options.ret_instruction_bytes.empty());
+    machine_code->insert(machine_code->end(), options.jmp_instruction_bytes.begin(),
+                         options.jmp_instruction_bytes.end());
+    machine_code->insert(machine_code->end(), options.ret_instruction_bytes.begin(),
+                         options.ret_instruction_bytes.end());
+  }
+
+ protected:
+  PeCoffEpilog pe_coff_epilog_;
+  std::unique_ptr<MemoryFake> process_mem_fake_;
+  // Anything we do in the tests will increase the stack pointer value, so this is a safe starting
+  // point.
+  uint64_t expected_stack_pointer_after_unwind_ = 0;
+};
+
+TEST_F(PeCoffEpilogTest, fails_with_error_if_disassembling_fails) {
+  std::vector<uint8_t> machine_code{0x48, 0x81, 0xc1, 0x07, 0xc3};
```

Comment:
Not sure if this is really feasible, but consider adding constants (or macros) for the bytes that you use here, to make it more readable.

Reply:
Makes sense. I think with the comments is fine.